tests/test_sft_trainer.py

```diff
@@ -423,6 +423,25 @@ def test_data_collator_completion_lm(self):
         result_text = self.tokenizer.decode(batch["input_ids"][0, last_pad_idx + 1 :])
         self.assertTrue(result_text == "I have not been masked correctly.")
 
+    def test_data_collator_chat_completion_lm(self):
+        human_template = "### Human:"
+        assistant_template = "### Assistant:"
+        data_collator = DataCollatorForChatCompletionOnlyLM(
+            human_template, assistant_template, tokenizer=self.tokenizer, mlm=False
+        )
+
+        text = """### Human: Hello all this should be masked.### Assistant: I have not been masked correctly.### Human: All this should be masked too.### Assistant: I have not been masked correctly too."""
+        encoded_text = self.tokenizer(text)
+        encoded_text["input_ids"] = encoded_text["input_ids"]
+
+        examples = [encoded_text]
+
+        batch = data_collator(examples)
+        labels = batch["labels"]
+        non_masked_tokens = batch["input_ids"][labels != -100]
+        result_text = self.tokenizer.decode(non_masked_tokens)
+        self.assertTrue(result_text == " I have not been masked correctly.### I have not been masked correctly too.")
```

Comment:
It is a bit surprising that there is `###` in the non-masked part. I think adding a space or a line break between dots and `###` would solve the issue. Can you try that out and see if it works? 🙏

Reply:
The problem with this is that if we do not train the model to output the human response token (either an added special token or the first token of it), the model will just spit out text until max_length. We need a way to stop the model from generating text once it reaches a certain token (I have chosen the human response tokens but it is also possible to add an eos token at the end of each assistant response). What do you think? (sorry if I was unclear)

Reply:
Basically, the idea is to help the model know that this is the end of the assistant response.

Reply:
This seems like a fundamental problem and I see both of your opinions. Ideally, every turn ends with an `endofturn` token. I believe that the Chat UI does this. It adds a `</s>` token [at the end of each turn](https://github.com/huggingface/chat-ui/blob/569bde33470b075bf1365af2cb03a1b31b875379/src/lib/server/websearch/generateQuery.ts#L9). Llama 2 uses some special tokens to indicate end/start of user queries vs system queries (https://github.com/facebookresearch/llama/blob/6c7fe276574e78057f917549435a2554000a876d/llama/generation.py#L44-L45). So the question here becomes how we can accomodate for that I think...